### DISCUSSION TITLE
Disabling new org creation in self hosted multi-tenant setup.

### DISCUSSION THREAD

**Discussion by fabiankaestner** [Budibase/budibase#3044] [Ideas]
It should be possible to disable the creation of new orgs in a self hosted multi-tenant setup. Otherwise anybody on the internet can come along and create new applications on your system, unless it sits inside a VPN or an authenticating reverse proxy.

@shogunpurple

**Comment by PatrickHuetter** [comment]
I've seen a configuration param in values.yaml for that:
```  
multiTenancy: "0" # set to 0 to disable multiple orgs, set to 1 to enable multiple orgs
```

Maybe it helps? Does budibase work for you with the current helm chart? I don't get it working yet.

**Reply by fabiankaestner** [reply]
Yes, funnily enough I wrote the PR that added this earlier 😄 .

This parameter, as I understand it, only controls if budibase is setup to handle multiple organizations or not. What I am proposing is having a multi-tenant setup with multiple organizations, and then disabling the setup of any further organizations, which I don't think is how this parameter currently works.

As for your issue getting the chart to work, are you working with the actual chart from the repository or the chart currently in the budibase helm repo. I don't know if the current chart has been pushed to the helm repo yet, as the latest PR was merged earlier today.

I have the chart working on two kubernetes clusters, with an issue we haven't figured out yet on a third cluster.

**Reply by PatrickHuetter** [reply]
Yeah that's really funny 😄.
Okay that's also a feature that i would like to use.

I'm using the current stable from helm repo but it seems that this one isn't really working. I can't create the organization and admin user. If you could share some parts of your helm values config that's working i would be glad. I could also switch to the helm chart version from this git repo here.

My problem is described here:
https://github.com/Budibase/budibase/discussions/3021#discussioncomment-1494644

**Reply by shogunpurple** [reply]
Hey guys - I've rolled out the latest helm chart to master now - if you pull the latest from helm it will have the latest changes.

**Reply by PatrickHuetter** [reply]
@shogunpurple Thanks i just tested the new version. I don't know wether it already contains the proposal of @fabiankaestner but it's also furthermore not working for me. I can't create the admin user successfully. All services are up and running, everything is green. Seems that the account gets also created with the red error toasty message but i can't login afterwards. I already tested every possible scenario with different helm variables but nothing works.

**Reply by shogunpurple** [reply]
@PatrickHuetter could you please provide the error messages that you see in the network tab when you try to sign up? Just trying to nail down exactly where it's going wrong.

**Reply by PatrickHuetter** [reply]
I got it! I checked every single line in the logs and did some tests. Seems to be a startup order issue of the pods in kubernetes. We have a very fast kubernetes cluster. In our case the worker AND app service are faster spawned then the redis service. That means they can't connect to the redis instance. After restarting BOTH services, worker and app after everything has spawned, lets the services connect to redis successfully and then everything works. So it would be helpful to have some init/health checks while startup dependend on redis. So that the health status or finished status of app and worker are only there, if startup was really successful. Maybe just configure the node server to quit if redis connection wasn't successful.

**Reply by PatrickHuetter** [reply]
/UPDATE/ Now i can't create apps because of:
```
message: "write EPROTO 140662116470088:error:1408F10B:SSL routines:ssl3_get_record:wrong version number:../deps/openssl/openssl/ssl/record/ssl3_record.c:332:\n"
status: 500
```

We have a clean kubernetes environment, fully working with cert-manager issuing valid letsencrypt certificates and using nginx ingress controller.

/UPDATE2/
Some apps get created and can be clicked in the apps overview then. So is "half" working.

/UPDATE3/
Now client-library.js is missing. But this seems to be an auth problem? I'm loggend in but some requests fail, not all.
![Bildschirmfoto 2021-10-20 um 12 50 49](https://user-images.githubusercontent.com/1142911/138080061-560747cc-b91b-4032-b520-ee204071e178.png)

**Reply by shogunpurple** [reply]
Thanks @PatrickHuetter - that's great feedback, I've actually not been able to reproduce that in my local clusters but if we add a health check or liveness probe to the pods, it should wait for redis to come up when using the internal one shipped with our helm chart.

@fabiankaestner - I think you might have also been experiencing the same issue here. I looked at the logs you sent, and it appears the worker cannot connect to the internal redis as well.

Ok, with the client library issue, where is it trying to fetch the client library from? It should be getting it from MinIO.

**Reply by PatrickHuetter** [reply]
I'm glad @shogunpurple. Just some hints for the health checks: I would strongly suggest to do them in a separate thread/worker every x seconds, e.g. every 10 seconds. Or as alternative use the kubernetes health checks via http (recommended) and let them run every 10s. Within this health check endpoint/call you should check every dependency. E.g. minio AND redis. If some of them is missing or has a bad state just don't return 200 http status and kubernetes will backoff restart the pod again and again until all dependencies are fullfilled. 

I'll next check minio...maybe there is also a problem with it.

**Reply by PatrickHuetter** [reply]
@shogunpurple The errors with ssl i posted before are also minio related. Seems to be everything related to the same problem. Might it be that the application wants to connect via SSL/https (internally in k8s networking/namespace) instead of using http? Also wether to use or not use ssl should be configurable if it's not yet. The default for kubernetes internal networking should be http in my opinion since most kubernetes clusters youse private LAN or have wireguard encrypted traffic for their SDN (software defined networking).

```
ERROR [Wed Oct 20 2021 11:09:47 GMT+0000 (Coordinated Universal Time)] (28 on minio-service): write EPROTO 140662116470088:error:1408F10B:SSL routines:ssl3_get_record:wrong version number:../deps/openssl/openssl/ssl/record/ssl3_record.c:332:
--
Wed, Oct 20 2021 1:09:47 pm |  
Wed, Oct 20 2021 1:09:47 pm | Error: write EPROTO 140662116470088:error:1408F10B:SSL routines:ssl3_get_record:wrong version number:../deps/openssl/openssl/ssl/record/ssl3_record.c:332:
Wed, Oct 20 2021 1:09:47 pm |  
Wed, Oct 20 2021 1:09:47 pm | at WriteWrap.onWriteComplete [as oncomplete] (internal/stream_base_commons.js:94:16)
Wed, Oct 20 2021 1:09:47 pm | at WriteWrap.callbackTrampoline (internal/async_hooks.js:126:14)
Wed, Oct 20 2021 1:09:47 pm | Trace: Error: write EPROTO 140662116470088:error:1408F10B:SSL routines:ssl3_get_record:wrong version number:../deps/openssl/openssl/ssl/record/ssl3_record.c:332:
Wed, Oct 20 2021 1:09:47 pm |  
Wed, Oct 20 2021 1:09:47 pm | at WriteWrap.onWriteComplete [as oncomplete] (internal/stream_base_commons.js:94:16)
Wed, Oct 20 2021 1:09:47 pm | at WriteWrap.callbackTrampoline (internal/async_hooks.js:126:14) {
Wed, Oct 20 2021 1:09:47 pm | errno: 'EPROTO',
Wed, Oct 20 2021 1:09:47 pm | code: 'NetworkingError',
Wed, Oct 20 2021 1:09:47 pm | syscall: 'write',
Wed, Oct 20 2021 1:09:47 pm | region: 'us-east-1',
Wed, Oct 20 2021 1:09:47 pm | hostname: 'minio-service',
Wed, Oct 20 2021 1:09:47 pm | retryable: true,
Wed, Oct 20 2021 1:09:47 pm | time: 2021-10-20T11:09:47.168Z
Wed, Oct 20 2021 1:09:47 pm | }
Wed, Oct 20 2021 1:09:47 pm | at /app/dist/api/index.js:69:21
Wed, Oct 20 2021 1:09:47 pm | at Generator.throw (<anonymous>)
Wed, Oct 20 2021 1:09:47 pm | at rejected (/app/dist/api/index.js:6:65)
Wed, Oct 20 2021 1:09:47 pm | at runMicrotasks (<anonymous>)
Wed, Oct 20 2021 1:09:47 pm | at processTicksAndRejections (internal/process/task_queues.js:97:5)
Wed, Oct 20 2021 1:09:47 pm | INFO [1634728187173] (28 on app-service-77cdc556f-qwsq5): request errored
Wed, Oct 20 2021 1:09:47 pm | res: {
Wed, Oct 20 2021 1:09:47 pm | "statusCode": 500,
Wed, Oct 20 2021 1:09:47 pm | "headers": {
Wed, Oct 20 2021 1:09:47 pm | "vary": "Accept-Encoding",
Wed, Oct 20 2021 1:09:47 pm | "content-type": "application/json; charset=utf-8",
Wed, Oct 20 2021 1:09:47 pm | "content-length": "176"
Wed, Oct 20 2021 1:09:47 pm | }
Wed, Oct 20 2021 1:09:47 pm | }
Wed, Oct 20 2021 1:09:47 pm | err: {
Wed, Oct 20 2021 1:09:47 pm | "type": "Error",
Wed, Oct 20 2021 1:09:47 pm | "message": "failed with status code 500",
Wed, Oct 20 2021 1:09:47 pm | "stack":
Wed, Oct 20 2021 1:09:47 pm | Error: failed with status code 500
Wed, Oct 20 2021 1:09:47 pm | at ServerResponse.onResFinished (/app/node_modules/pino-http/logger.js:74:38)
Wed, Oct 20 2021 1:09:47 pm | at ServerResponse.emit (events.js:326:22)
Wed, Oct 20 2021 1:09:47 pm | at ServerResponse.EventEmitter.emit (domain.js:483:12)
Wed, Oct 20 2021 1:09:47 pm | at ServerResponse.emitted (/app/node_modules/emitter-listener/listener.js:134:21)
Wed, Oct 20 2021 1:09:47 pm | at onFinish (_http_outgoing.js:735:10)
Wed, Oct 20 2021 1:09:47 pm | at onCorkedFinish (_stream_writable.js:673:5)
Wed, Oct 20 2021 1:09:47 pm | at afterWrite (_stream_writable.js:490:5)
Wed, Oct 20 2021 1:09:47 pm | at afterWriteTick (_stream_writable.js:477:10)
Wed, Oct 20 2021 1:09:47 pm | at processTicksAndRejections (internal/process/task_queues.js:83:21)
Wed, Oct 20 2021 1:09:47 pm | }
Wed, Oct 20 2021 1:09:47 pm | responseTime: 924
Wed, Oct 20 2021 1:09:47 pm | req: {
Wed, Oct 20 2021 1:09:47 pm | "id": 186,
Wed, Oct 20 2021 1:09:47 pm | "method": "GET",
Wed, Oct 20 2021 1:09:47 pm | "url": "/api/app_dev_default_00b0adf2599346278f23ad55555a2883/components/definitions",
Wed, Oct 20 2021 1:09:47 pm | "headers": {
Wed, Oct 20 2021 1:09:47 pm | "host": "budibase.k8s.test.io",
Wed, Oct 20 2021 1:09:47 pm | "x-request-id": "25b4ad8e08883c336389214252a76444",
Wed, Oct 20 2021 1:09:47 pm | "x-real-ip": "46.87.155.242",
Wed, Oct 20 2021 1:09:47 pm | "x-forwarded-for": "46.87.155.242",
Wed, Oct 20 2021 1:09:47 pm | "x-forwarded-host": "budibase.k8s.test.io",
Wed, Oct 20 2021 1:09:47 pm | "x-forwarded-port": "443",
Wed, Oct 20 2021 1:09:47 pm | "x-forwarded-proto": "https",
Wed, Oct 20 2021 1:09:47 pm | "x-forwarded-scheme": "https",
Wed, Oct 20 2021 1:09:47 pm | "x-scheme": "https",
Wed, Oct 20 2021 1:09:47 pm | "pragma": "no-cache",
Wed, Oct 20 2021 1:09:47 pm | "cache-control": "no-cache",
Wed, Oct 20 2021 1:09:47 pm | "sec-ch-ua": "\"Chromium\";v=\"94\", \"Google Chrome\";v=\"94\", \";Not A Brand\";v=\"99\"",
Wed, Oct 20 2021 1:09:47 pm | "x-budibase-app-id": "",
Wed, Oct 20 2021 1:09:47 pm | "content-type": "application/json",
Wed, Oct 20 2021 1:09:47 pm | "sec-ch-ua-mobile": "?0",
Wed, Oct 20 2021 1:09:47 pm | "user-agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/94.0.4606.81 Safari/537.36",
Wed, Oct 20 2021 1:09:47 pm | "sec-ch-ua-platform": "\"macOS\"",
Wed, Oct 20 2021 1:09:47 pm | "accept": "*/*",
Wed, Oct 20 2021 1:09:47 pm | "sec-fetch-site": "same-origin",
Wed, Oct 20 2021 1:09:47 pm | "sec-fetch-mode": "cors",
Wed, Oct 20 2021 1:09:47 pm | "sec-fetch-dest": "empty",
Wed, Oct 20 2021 1:09:47 pm | "referer": "https://budibase.k8s.test.io/builder/app/app_dev_default_00b0adf2599346278f23ad55555a2883/design/screen/screen_cbb9589fb7da43e483c4e334ba7c996d",
Wed, Oct 20 2021 1:09:47 pm | "accept-encoding": "gzip, deflate, br",
Wed, Oct 20 2021 1:09:47 pm | "accept-language": "de-DE,de;q=0.9",
Wed, Oct 20 2021 1:09:47 pm | "cookie": "budibase:auth=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJ1c2VySWQiOiJ1c19kMTViYzRmMTU4YjI0MGE1OGQ4ZWE2YWRlMDY0OWFhMiIsInNlc3Npb25JZCI6ImY4ZDliMDY0NTNhMDQ2ZTY5NTVjMDAxNjJlYmE2YzZkIiwidGVuYW50SWQiOiJkZWZhdWx0IiwiaWF0IjoxNjM0NzI2NDc2fQ.qsN3jzy7RT7pRwaeGsNFYvfMjX0fxAkojCUQbQ6-HNs; budibase:auth.sig=a3s1Rj4r4Gn-Zifaj-75Ly253bM; budibase:currentapp=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJhcHBJZCI6ImFwcF9kZXZfZGVmYXVsdF8wMGIwYWRmMjU5OTM0NjI3OGYyM2FkNTU1NTVhMjg4MyIsImlhdCI6MTYzNDcyNjk0OX0.L803re8u26OETZYpGr-CElAmqg49xsYmCdJzSwezPTY",
Wed, Oct 20 2021 1:09:47 pm | "x-envoy-expected-rq-timeout-ms": "15000",
Wed, Oct 20 2021 1:09:47 pm | "content-length": "0"
Wed, Oct 20 2021 1:09:47 pm | },
Wed, Oct 20 2021 1:09:47 pm | "remoteAddress": "::ffff:10.42.1.154",
Wed, Oct 20 2021 1:09:47 pm | "remotePort": 54250
Wed, Oct 20 2021 1:09:47 pm | }
Wed, Oct 20 2021 1:09:47 pm | INFO [1634728187370] (28 on app-service-77cdc556f-qwsq5): request completed
Wed, Oct 20 2021 1:09:47 pm | res: {
Wed, Oct 20 2021 1:09:47 pm | "statusCode": 200,
Wed, Oct 20 2021 1:09:47 pm | "headers": {
Wed, Oct 20 2021 1:09:47 pm | "vary": "Accept-Encoding",
Wed, Oct 20 2021 1:09:47 pm | "content-type": "application/json; charset=utf-8",
Wed, Oct 20 2021 1:09:47 pm | "content-length": "36"
Wed, Oct 20 2021 1:09:47 pm | }
Wed, Oct 20 2021 1:09:47 pm | }
```

**Reply by shogunpurple** [reply]
@PatrickHuetter thanks for that, we do some `livenessProbe` stuff in other parts of the K8S templates, but that would be a good one to ensure cluster health for clusters that are spinning up those services very quickly.

Ok, so it looks like the real MinIO URL is not being set and your infra is trying to go to s3. In your `values.yaml`, set `services.objectStore.url` to:

`minio-service:9000`

and update your helm chart. Does that solve your issue?

**Reply by shogunpurple** [reply]
I've created a PR to default `MINIO_URL` to that for both worker and app-service when it's not set in the `values.yaml`.

https://github.com/Budibase/budibase/pull/3099

Once merged you should be able to just pull the latest helm chart and get the above functionality.

**Reply by PatrickHuetter** [reply]
@shogunpurple just to let you know: I did the changes you mentioned and also made a whole re-install. Same problem exists furthermore.

**Reply by shogunpurple** [reply]
Ok, so is it still trying to go to S3?

Would you be able to show me your `values.yaml` without sensitive data? That would really help to debug.

**Reply by PatrickHuetter** [reply]
Yes no problem. My current values.yaml is the following (complete):
```
    ingress:
      enabled: true
      nginx: false
      className: ""
      annotations:
        kubernetes.io/ingress.class: nginx
        cert-manager.io/cluster-issuer: letsencrypt-production
      hosts:
        - host: budibase.somedomain.io
          paths:
            - path: /
              pathType: Prefix
              backend:
                service:
                  name: proxy-service
                  port:
                    number: 10000
      tls:
        - hosts:
            - budibase.somedomain.io
          secretName: budibase-cert-tls
    #globals:
      #enableAnalytics: false
      #selfHosted: 1
    services:
      objectStore:
        url: "minio-service:9000"
    #selfHosted: 1
```

**Reply by shogunpurple** [reply]
It appears your `globals.selfHosted` variable has been commented out - this needs to be `1` - I think that could be why your server is trying to go to s3.

**Reply by PatrickHuetter** [reply]
@shogunpurple I did everything you said and problem still exists (globals.selfHosted = 1). Also the creation of apps doesn't work really.
<img width="1306" alt="Bildschirmfoto 2021-10-20 um 13 46 31" src="https://user-images.githubusercontent.com/1142911/138090214-79e1e301-592f-42bd-aef4-5438f5c893b3.png">

**Reply by fabiankaestner** [reply]
> I've created a PR to default `MINIO_URL` to that for both worker and app-service when it's not set in the `values.yaml`.
> 
> https://github.com/Budibase/budibase/pull/3099
> 
> Once merged you should be able to just pull the latest helm chart and get the above functionality.

This should have been already included in my PR, did I miss something?

**Reply by shogunpurple** [reply]
@PatrickHuetter I think this is possibly something to do with the SSL setup - the server is forwarding a request on port 443 when it should be straight up HTTP.

Could I see your worker and server logs when you try to create an app?

It also seems like your `values.yaml` is heavily modified - is MinIO definitely up? `minio: true` is missing from the `values.yaml` and all the redis config as well. 

@fabiankaestner nope, it was fine - your changes just needed merged into master, which I have done.

**Comment by AiCaptain** [comment]
Hi guys, 
are there any plans to implement the behavior @fabiankaestner described in the initial post? 

I think the multitenancy is a nice feature but the creation of tenants should be under control of the admin(s) and in the moment it's a little bit error prone that a typo will lead to new tenant ...

Thank you in advance,
André

**Reply by shogunpurple** [reply]
@AiCaptain full multitenancy will be an enterprise feature in future - in its current form it isn’t really meant to be used in self hosted installations and it’s just an escape hatch for development
